File: framework/extracted/Ahorn/entities/MorphoKnightClone.jl
module MorphoKnightClone

using ..Ahorn, Maple

@mapdef Entity "MaggyHelper/MorphoKnightClone" MorphoKnightClone(x::Integer, y::Integer, attackCooldown::Number=1.05, bossMusic::String="event:/pusheen/music/lvl15/morpho_knight", chargeAnimationPath::String="vortex_summon", dashSpeed::Number=320.0, health::Integer=10, idleAnimationPath::String="swords", moveAnimationPath::String="swords", moveSpeed::Number=142.0, orbitRadius::Number=66.0, playMusicOnStart::Bool=false, slashAnimationPath::String="double_side_slash", spritePath::String="characters/els_true_final_boss/Maggy/DesoloZantas/siamo_zero_morpho_knight_fake/", warpAnimationPath::String="vortex_strike")

const placements = Ahorn.PlacementDict(
    "default" => Ahorn.EntityPlacement(MorphoKnightClone)
)

function Ahorn.selection(entity::MorphoKnightClone)
    return Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::MorphoKnightClone, room::Maple.Room)
    Ahorn.drawRectangle(ctx, Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16), (0.5, 0.5, 0.5, 0.5), (0.0, 0.0, 0.0, 1.0))
end

end

File: framework/extracted/Loenn/entities/morpho_knight_clone.lua
local drawableSprite = require("structs.drawable_sprite")
local utils = require("utils")

local morphoKnightClone = {}

morphoKnightClone.name = "MaggyHelper/MorphoKnightClone"
morphoKnightClone.depth = -8600
morphoKnightClone.justification = {0.5, 1.0}

morphoKnightClone.placements = {
    {
        name = "default",
        data = {
            health = 10,
            moveSpeed = 142.0,
            attackCooldown = 1.05,
            orbitRadius = 66.0,
            dashSpeed = 320.0,
            playMusicOnStart = false,
            bossMusic = "event:/music/pusheen/lvl15/morpho_knight",
            spritePath = "characters/els_true_final_boss/siamo_zero_morpho_knight_fake/",
            idleAnimationPath = "swords",
            moveAnimationPath = "swords",
            chargeAnimationPath = "vortex_summon",
            slashAnimationPath = "double_side_slash",
            warpAnimationPath = "vortex_strike"
        }
    }
}

morphoKnightClone.fieldInformation = {
    health = {
        fieldType = "integer",
        minimumValue = 1
    },
    moveSpeed = {
        minimumValue = 1.0
    },
    attackCooldown = {
        minimumValue = 0.1
    },
    orbitRadius = {
        minimumValue = 8.0
    },
    dashSpeed = {
        minimumValue = 10.0
    },
    playMusicOnStart = {
        fieldType = "boolean"
    },
    bossMusic = {
        fieldType = "string",
        editable = true
    },
    spritePath = {
        fieldType = "string",
        editable = true
    },
    idleAnimationPath = {
        fieldType = "string",
        editable = true
    },
    moveAnimationPath = {
        fieldType = "string",
        editable = true
    },
    chargeAnimationPath = {
        fieldType = "string",
        editable = true
    },
    slashAnimationPath = {
        fieldType = "string",
        editable = true
    },
    warpAnimationPath = {
        fieldType = "string",
        editable = true
    }
}

morphoKnightClone.fieldOrder = {
    "x", "y",
    "health",
    "moveSpeed",
    "attackCooldown",
    "orbitRadius",
    "dashSpeed",
    "playMusicOnStart",
    "bossMusic",
    "spritePath",
    "idleAnimationPath",
    "moveAnimationPath",
    "chargeAnimationPath",
    "slashAnimationPath",
    "warpAnimationPath"
}

local fallbackPreviewTextures = {
    "characters/els_true_final_boss/siamo_zero_morpho_knight_fake/swords00",
    "objects/bosses/metaKnightBoss/idle00"
}

local function combineTexturePath(root, animationPath)
    if not root or root == "" or not animationPath or animationPath == "" then
        return nil
    end

    local normalizedRoot = root:gsub("/+$", "")
    local normalizedAnimation = animationPath:gsub("^/+", "")
    return normalizedRoot .. "/" .. normalizedAnimation .. "00"
end

local function getSprite(entity)
    local candidateTextures = {
        combineTexturePath(entity.spritePath, entity.idleAnimationPath)
    }

    for _, texture in ipairs(fallbackPreviewTextures) do
        table.insert(candidateTextures, texture)
    end

    for _, texture in ipairs(candidateTextures) do
        if texture then
        local ok, sprite = pcall(drawableSprite.fromTexture, texture, entity)
        if ok and sprite then
            sprite:setJustification(0.5, 1.0)
            sprite:setColor({1.0, 0.67, 0.42, 1.0})
            return sprite
        end
        end
    end

    return nil
end

function morphoKnightClone.sprite(room, entity)
    local sprite = getSprite(entity)
    return sprite and {sprite} or {}
end

function morphoKnightClone.selection(room, entity)
    return utils.rectangle(entity.x - 16, entity.y - 36, 32, 40)
end

return morphoKnightClone
